File: phase2/backend/src/api/tasks.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlmodel import Session
from phase2.backend.src.db.session import get_session
from phase2.backend.src.models import Task, User
from phase2.backend.src.schemas.task import TaskCreate, TaskUpdate, ReminderCreate
from phase2.backend.src.services.reminder_service import ReminderService
from phase2.backend.src.services.task_crud_service import TaskCRUDService
from phase2.backend.src.services.task_query_service import TaskQueryService
from phase2.backend.src.api.auth import get_current_user
from phase2.backend.src.models import User

def create_tasks_router() -> APIRouter:
    router = APIRouter(prefix="/tasks", tags=["tasks"])

    @router.post("/", response_model=Task, status_code=status.HTTP_201_CREATED)
    def create_task(
        task_create: TaskCreate,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Create a new task.
        """
        task = Task.from_orm(task_create)
        task.user_id = current_user.id
        crud_service = TaskCRUDService(session)
        return crud_service.create_task(task)

    @router.get("/", response_model=List[Task])
    def list_tasks(
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session),
        search: Optional[str] = Query(None, description="Search tasks by title or description"),
        priority: Optional[str] = Query(None, description="Filter by priority (low, medium, high)"),
        status: Optional[str] = Query(None, description="Filter by status (pending, in progress, completed, archived, cancelled)"),
        tags: Optional[List[str]] = Query(None, description="Filter by tags"),
        sort_by: Optional[str] = Query("due_date", description="Sort by field (priority, due_date, alpha)"),
        sort_order: Optional[str] = Query("asc", description="Sort order (asc, desc)")
    ):
        """
        List all tasks for the current user, with optional filtering, searching, and sorting.
        """
        query_service = TaskQueryService(session)
        return query_service.query_tasks(
            user_id=current_user.id,
            search=search,
            priority=priority,
            status=status,
            tags=tags,
            sort_by=sort_by,
            sort_order=sort_order,
        )

    @router.get("/{task_id}", response_model=Task)
    def get_task(
        task_id: int,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Get a specific task by its ID.
        """
        crud_service = TaskCRUDService(session)
        task = crud_service.get_task(task_id)
        if not task or task.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
        return task

    @router.put("/{task_id}", response_model=Task)
    def update_task(
        task_id: int,
        task_update: TaskUpdate,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Update a specific task by its ID.
        """
        crud_service = TaskCRUDService(session)
        task = crud_service.get_task(task_id)
        if not task or task.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
        
        updated_task = crud_service.update_task(task_id, task_update)
        return updated_task

    @router.delete("/{task_id}", status_code=status.HTTP_204_NO_CONTENT)
    def delete_task(
        task_id: int,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Delete a specific task by its ID.
        """
        crud_service = TaskCRUDService(session)
        task = crud_service.get_task(task_id)
        if not task or task.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
        
        if not crud_service.delete_task(task_id):
            raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail="Failed to delete task")

    @router.patch("/{task_id}/complete", response_model=Task)
    def complete_task(
        task_id: int,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Mark a specific task as complete.
        """
        crud_service = TaskCRUDService(session)
        task = crud_service.get_task(task_id)
        if not task or task.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")
        
        task.status = "completed"
        session.add(task)
        session.commit()
        session.refresh(task)
        return task

    @router.post("/{task_id}/reminders", status_code=status.HTTP_201_CREATED)
    def create_reminder(
        task_id: int,
        reminder_create: ReminderCreate,
        current_user: User = Depends(get_current_user),
        session: Session = Depends(get_session)
    ):
        """
        Create a reminder for a specific task.
        """
        crud_service = TaskCRUDService(session)
        task = crud_service.get_task(task_id)
        if not task or task.user_id != current_user.id:
            raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Task not found")

        reminder_service = ReminderService(session)
        reminder_service.create_reminder(task, reminder_create.reminder_date)
        return {"message": "Reminder created successfully"}
    
    return router

router = create_tasks_router()

File: phase5/frontend/src/lib/api/tasks.ts
import apiClient from '@/lib/api-client';
import { Task } from '@/types/task';
import { normalizeTask, normalizeTasks } from '@/lib/task-utils';

// Define TypeScript interfaces matching the backend models
export interface TaskCreateData {
  title: string;
  description?: string;
  priority?: string; // low, medium, high, urgent
  due_date?: string; // ISO date string
  reminder_time?: string; // ISO date string for reminder
  tags?: string; // comma-separated tags
  recurrence_pattern?: RecurrencePattern; // Recurrence pattern for recurring tasks
  parent_task_id?: string; // UUID as string for hierarchical tasks
  status?: string; // pending, in progress, completed, archived, cancelled
}

export interface TaskUpdateData {
  title?: string;
  description?: string;
  completed?: boolean;
  priority?: string; // low, medium, high, urgent
  due_date?: string; // ISO date string
  reminder_time?: string; // ISO date string for reminder
  tags?: string; // comma-separated tags
  recurrence_pattern?: RecurrencePattern; // Recurrence pattern for recurring tasks
  parent_task_id?: string; // UUID as string for hierarchical tasks
  status?: string; // pending, in progress, completed, archived, cancelled
}

export interface RecurrencePattern {
  frequency: 'daily' | 'weekly' | 'monthly' | 'yearly' | 'custom'; // Recurrence frequency
  interval: number; // Interval multiplier (e.g., every 2 weeks)
  end_condition: {
    type: 'never' | 'on_date' | 'after_occurrences'; // When to stop recurrence
    value?: string | number; // Date string or occurrence count
  };
  exceptions?: string[]; // Array of dates to skip in recurrence
}

// API functions for task operations
// Note: These functions use the authenticated user from the JWT token
export const taskApi = {
  // Create a new task
  createTask: async (userId: string, taskData: TaskCreateData): Promise<Task> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.post<Task>(`/api/${userId}/tasks`, taskData);
    return normalizeTask(response.data);
  },

  // Get all tasks for the authenticated user with optional filtering
  getTasks: async (
    userId: string,
    search?: string,
    priority?: string,
    status?: string,
    tags?: string[],
    sortBy: string = "due_date",
    sortOrder: string = "asc"): Promise<Task[]> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    // Build query parameters
    const params = new URLSearchParams();
    if (search) params.append('search', search);
    if (priority) params.append('priority', priority);
    if (status) params.append('status', status);
    if (tags) tags.forEach(tag => params.append('tags', tag));
    params.append('sort_by', sortBy);
    params.append('sort_order', sortOrder);

    const queryString = params.toString();
    const url = `/api/${userId}/tasks${queryString ? '?' + queryString : ''}`;

    const response = await apiClient.get<Task[]>(url);
    return normalizeTasks(response.data);
  },

  // Get a specific task by ID
  getTaskById: async (userId: string, taskId: string): Promise<Task> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.get<Task>(`/api/${userId}/tasks/${taskId}`);
    return normalizeTask(response.data);
  },

  // Update a task
  updateTask: async (userId: string, taskId: string, taskData: TaskUpdateData): Promise<Task> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.put<Task>(`/api/${userId}/tasks/${taskId}`, taskData);
    return normalizeTask(response.data);
  },

  // Delete a task
  deleteTask: async (userId: string, taskId: string): Promise<void> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    await apiClient.delete(`/api/${userId}/tasks/${taskId}`);
  },

  // Mark a task as complete
  completeTask: async (userId: string, taskId: string): Promise<Task> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.patch<Task>(`/api/${userId}/tasks/${taskId}/complete`);
    return normalizeTask(response.data);
  },

  // Get child tasks for a parent task
  getChildTasks: async (userId: string, taskId: string): Promise<Task[]> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.get<Task[]>(`/api/${userId}/tasks/${taskId}/children`);
    return normalizeTasks(response.data);
  },

  // Create a child task under a parent task
  createChildTask: async (userId: string, taskId: string, taskData: TaskCreateData): Promise<Task> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.post<Task>(`/api/${userId}/tasks/${taskId}/children`, taskData);
    return normalizeTask(response.data);
  },

  // Get parent tasks (ancestors) for a task
  getParentTasks: async (userId: string, taskId: string): Promise<Task[]> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.get<Task[]>(`/api/${userId}/tasks/${taskId}/ancestors`);
    return normalizeTasks(response.data);
  },

  // Get upcoming reminders for a user
  getUpcomingReminders: async (userId: string, hoursAhead: number = 24): Promise<any[]> => {
    if (!userId) {
      throw new Error('User ID is required');
    }

    const response = await apiClient.get<any[]>(`/api/${userId}/tasks/reminders/upcoming?hours_ahead=${hoursAhead}`);
    return response.data;
  },
};